XR_APILAYER_NOVENDOR_toolkit/d3d12.cpp

```diff
@@ -454,8 +457,22 @@ namespace {
             }
         }
 
-        void saveToFile(const std::string& path) const override {
-            // TODO: Implement this.
+        void saveToFile(const std::filesystem::path& path) const override {
+            const auto& fileFormat = path.extension() == ".png"   ? GUID_ContainerFormatPng
+                                     : path.extension() == ".bmp" ? GUID_ContainerFormatBmp
+                                     : path.extension() == ".jpg" ? GUID_ContainerFormatJpeg
+                                                                  : GUID_ContainerFormatDds;
+
+            ComPtr<ID3D12CommandQueue> commandQueue;
+            UINT size = sizeof(ID3D12CommandQueue*);
+            m_device->getNative<D3D12>()->GetPrivateData(IID_ID3D12CommandQueue, &size, set(commandQueue));
```

Comment:
This doesn't behave as intended I think.

You don't write directly to the command queue with D3D12,  instead you use a ID3D12CommandList, which you eventually execute via the command queue. My guess is that SaveeWICTextureToFile() just creates its own command list and submits it right away. 

When this saveToFile() method is invoked, you will have some pending command lists that are not committed to the command queue yet. It is that call to flushContext(), which occurs after takeScreenshot(), that commits the command list to the queue:

https://github.com/mbucchia/OpenXR-Toolkit/blob/38e4eb92546e99ba0669a18c037ef644df4e6a29/XR_APILAYER_NOVENDOR_toolkit/layer.cpp#L1561

Therefore, the texture has not been written to yet, and the screenshot commands are enqueued right away by the function below, which effectively captures the old content of the texture, 3 frames ago.

The solution I believe, is to add a call to flushContext() prior to taking the screenshot.

Reply:
I will investigate this, thanks!

Reply:
The implementation goes through [CaptureTexture](https://github.com/mbucchia/OpenXR-Toolkit/pull/219/commits/5cd1f2c9eb0ab5339062904c732f74fa35550049#diff-c09b71bc4901b5f7a9df2e0d99f7b40a3bf5dfdc42d0240151f81c54ad42dce3R835) which creates a command list, execute the list on the queue, and busy-wait on a fence until completion. The resource is captured and stored in a a resource until being written later on via `encoder->Commit()`. This seems ok to me.

Reply:
This actually illustrates the problem. 

You have the following happening

1) xrEndFrame begins writing to a Command-line
2) upscaling/post process/menu render commands are added to the Command list
3) take screenshot is called, it creates its own Command list, enqueue and wait for it
4) xrEndFrame is about to end, we take the Command list from 1 and 2) and we enqueue it

At step 3), the outcome of step 2) isn't written to the texture yet, because the commands are only enqueued at step 4).